apollo-federation/src/sources/connect/expand/mod.rs

```diff
@@ -799,36 +797,53 @@ mod helpers {
     fn extract_params_from_body(
         connector: &Connector,
     ) -> Result<HashSet<Parameter>, FederationError> {
-        let body_parameters = connector
-            .transport
-            .body
-            .as_ref()
-            .and_then(JSONSelection::extract_parameters)
-            .unwrap_or_default();
-
-        body_parameters
-            .iter()
-            .map(|StaticParameter { name, paths }| {
-                let mut parts = paths.iter();
-                let field = parts.next().ok_or(FederationError::internal(
-                    "expected parameter in JSONSelection to contain a field",
-                ))?;
-
-                match *name {
-                    "$args" => Ok(Parameter::Argument {
-                        argument: field,
-                        paths: parts.copied().collect(),
-                    }),
-                    "$this" => Ok(Parameter::Sibling {
-                        field,
-                        paths: parts.copied().collect(),
-                    }),
-                    other => Err(FederationError::internal(format!(
-                        "got unsupported parameter: {other}"
-                    ))),
-                }
-            })
-            .collect::<Result<HashSet<_>, _>>()
+        if let Some(body) = &connector.transport.body {
+            use crate::sources::connect::json_selection::CollectVarPaths;
+            let var_paths = body.collect_var_paths();
+
+            let mut results =
+                HashSet::with_capacity_and_hasher(var_paths.len(), Default::default());
+
+            for var_path in var_paths {
+                match var_path.var_name_and_nested_keys() {
+                    Some(("$args", keys)) => {
+                        let mut keys_iter = keys.into_iter();
+                        let first_key = keys_iter.next().ok_or(FederationError::internal(
+                            "expected at least one key in $args",
+                        ))?;
+                        results.insert(Parameter::Argument {
+                            argument: first_key,
+                            paths: keys_iter.collect(),
+                        });
+                    }
+                    Some(("$this", keys)) => {
+                        let mut keys_iter = keys.into_iter();
+                        let first_key = keys_iter.next().ok_or(FederationError::internal(
+                            "expected at least one key in $this",
+                        ))?;
+                        results.insert(Parameter::Sibling {
+                            field: first_key,
+                            paths: keys_iter.collect(),
+                        });
+                    }
+                    Some((other, _)) => {
+                        return Err(FederationError::internal(format!(
+                            "got unsupported parameter: {other}"
+                        )))
+                    }
```

Comment:
We actually _do_ support another var—it's `$config`, and we're likely to add more. I think it's a good idea for us to just create an enum for these that we start using everywhere, since we have them defined in a ton of places right now (and they're hard to find, speaking from experience when adding `$config`).

Reply:
Yeah I agree, as this will help us catch these missing implementations at compile time.

Reply:
Note that `$config` will need its own `Parameter::*` variant (not `Sibling` or `Argument`), which will need to be handled appropriately.

Is there a way to have a union of specific `&str` values, or does it have to be a symbolic set of named variants that we interconvert to/from `&str` strings?

Reply:
I added a TODO about adding/supporting `Parameter::Context` in https://github.com/apollographql/router/pull/5762/commits/483adf80d14772c6644c21d41b9e177116d125f3, for whoever wants to tackle that.

Reply:
It should be `Parameter::Config`, referring to `$config`. I'll add a Jira issue to track implementing it.

Reply:
`$config`: https://github.com/apollographql/router/pull/5762/commits/faada354f38e15a02096aa903ab93b856efa9e8a